persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,617 @@
+    , renderErrors
+    , runConfiguredParser
+    , initialExtraState
+    ) where
+
+import Control.Monad (void)
+import Control.Monad.Trans.State
+import Data.Either (partitionEithers)
+import Data.Foldable (fold)
+import Data.List (intercalate)
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NEL
+import qualified Data.Map as M
+import Data.Maybe
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Void
+import Database.Persist.Types
+import Database.Persist.Types.SourceSpan
+import Language.Haskell.TH.Syntax (Lift)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+-- We'll augment the parser with extra state to accumulate comments seen during parsing.
+-- Comments are lexed as whitespace, but will be used to generate documentation later.
+data ExtraState = ExtraState
+    { esPositionedCommentTokens :: [(SourcePos, CommentToken)]
+    , esLastDocumentablePosition :: Maybe SourcePos
+    }
+
+-- @since 2.16.1.0
+initialExtraState :: ExtraState
+initialExtraState =
+    ExtraState
+        { esPositionedCommentTokens = []
+        , esLastDocumentablePosition = Nothing
+        }
+
+type Parser a =
+    StateT
+        ExtraState
+        (Parsec Void String)
+        a
+
+type EntityParseError = ParseErrorBundle String Void
+type InternalParseResult a = Either EntityParseError (a, ExtraState)
+
+type ParseResult a = Either EntityParseError a
+type CumulativeParseResult a = Either [EntityParseError] a
+
+cumulativeData :: (Monoid a) => CumulativeParseResult a -> a
+cumulativeData cpr = case cpr of
+    Left _ -> mempty
+    Right r -> r
+
+-- | Run a parser using a provided ExtraState
+-- @since 2.16.0.0
+runConfiguredParser
+    :: ExtraState
+    -> Parser a
+    -> String
+    -> String
+    -> InternalParseResult a
+runConfiguredParser acc parser fp s = parseResult
+  where
+    (_internalState, parseResult) = runParser' (runStateT parser acc) initialInternalState
+    initialSourcePos =
+        SourcePos
+            { sourceName = fp
+            , sourceLine = pos1
+            , sourceColumn = pos1
+            }
+    initialPosState =
+        PosState
+            { pstateInput = s
+            , pstateOffset = 0
+            , pstateSourcePos = initialSourcePos
+            , -- for legacy compatibility, we treat each tab as a single unit of whitespace
+              pstateTabWidth = pos1
+            , pstateLinePrefix = ""
+            }
+    initialInternalState =
+        State
+            { stateInput = s
+            , stateOffset = 0
+            , statePosState = initialPosState
+            , stateParseErrors = []
+            }
+
+-- | Converts the errors in a CumulativeParseResult to a String
+-- @since 2.16.0.0
+renderErrors :: CumulativeParseResult a -> Maybe String
+renderErrors cpr = case cpr of
+    Right _ -> Nothing
+    Left errs -> Just $ intercalate "\n" $ fmap errorBundlePretty errs
```

Comment:
This function destroys information - you can see it syntactically in `Right _ -> Nothing` ! 

Instead of operating on a `Either Err a -> Maybe String`, we should do `Err -> String`. That way we are doing as little as possible with each function.

End users can recover the same behavior by using the `either` function to handle the `Either` - `either (Just . renderErrors) (const Nothing)` will work on `Either Err a -> Maybe String` type. But we only want users to be able to render errors if they actually *have* errors, which they'd acquire by way of:

```haskell
case parseResult of
  Left errs -> fail $ renderErrors errs
  Right a -> lift a
```